source/Personality.cpp

```diff
@@ -130,27 +130,32 @@ void Personality::Load(const DataNode &node)
 {
 	bool add = (node.Token(0) == "add");
 	bool remove = (node.Token(0) == "remove");
-	if(!(add || remove))
+	int keyIndex = add || remove;
+	if(!keyIndex)
 		flags.reset();
-	for(int i = 1 + (add || remove); i < node.Size(); ++i)
+	for(int i = 1 + keyIndex; i < node.Size(); ++i)
 		Parse(node, i, remove);
 
 	for(const DataNode &child : node)
 	{
 		if(child.Token(0) == "confusion")
 		{
-			if(add || remove)
+			if(keyIndex)
 				child.PrintTrace("Error: Cannot \"" + node.Token(0) + "\" a confusion value:");
 			else if(child.Size() < 2)
 				child.PrintTrace("Skipping \"confusion\" tag with no value specified:");
 			else
 				confusionMultiplier = child.Value(1);
 		}
+		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
+			lingerTime = -1;
+		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
+			lingerTime = child.Value(add + 1);
+		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
+			child.PrintTrace("Invalid \"linger time\" specification.");
```

Comment:
```c++
		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
			lingerTime = -1;
		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
			lingerTime = child.Value(add + 1);
		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
			child.PrintTrace("Invalid \"linger time\" specification.");
```

I'm not really sure what's going on here.
We're currently looking at the children of the `personality` node.
The parent `personality` node may take any of the following forms:
```
personality [flags...]
add personality [flags...]
remove personality [flags...]
```
The `flags` are any of the personalities. `confusion` is not allowed there, and neither is `"linger time"`. Not important here.
The bool `add` will be true if the first token in that parent is `add`, false otherwise. The bool `remove` will be true if that first token is `remove`, false otherwise.
Looking at the child nodes, we could have:
```
personality
	"linger time" <time#> # assigns the time to the linger time variable.
add personality
	"linger time" <time#> # assigns the time to the linger time variable.
remove personality
	"linger time" # sets the linger time variable to -1. Removing the linger time so no value needs to be provided.
```
The first branch of your code seems to deal with the `remove` case.
```c++
		else if(remove && child.Size() == 2 && child.Token(1) == "linger time")
			lingerTime = -1;
```
If checks if `remove` is true, ok. But then, it seems to expect the child to have two tokens, and the second of those to be `"linger time"`. This seems to expect the syntax:
```
	remove "linger time"
```
But, the `remove` bool you're checking was set based on the presence of `remove` on the `personality` node, not in this child.
If we want:
```
personality
	remove "linger time"
```
Instead of:
```
remove personality
	"linger time"
```
Then this code needs to be updated to check the first token of the child for `remove`, not reuse the check on the parent node.
Having `remove "linger time"` when we don't have `remove <flag>` seems like it'll be confusing, so I think we should stick to `remove` going on the parent node.
In that case, this first branch should be:
```c++
		else if(remove && child.Token(0) == "linger time")
			lingerTime = -1;
```
I don't think we need to consider this whole child node invalid if someone's done:
```
remove personality
	"linger time" 500
```
Or similar, but a warning about the additional token not having any effect might be useful:
```c++
		else if(remove && child.Token(0) == "linger time")
		{
			lingerTime = -1;
			if(child.Size() > 1)
				// Print a warning or something about the additional token(s) being ignored.
		}
```

```c++
		else if(!remove && child.Size() == 2 + keyIndex && child.Token(keyIndex) == "linger time")
			lingerTime = child.Value(add + 1);
```
The next branch assigns the given value to the `lingerTime` variable. But it's using the index variable that is based onthe parent node, not this node.
Say we had:
```
add personality
	"linger time" 500
```
`keyIndex` is going to be `1`, and `add` is going to be true.
So, this is expecting the child to have three tokens, and for `"linger time"` to be the second, and the value to be the third. Even if someone did:
```
personality
	add "linger time" 500
```
That still wouldn't work because `keyIndex` is now `0` and `add` is false, so it's expecting two tokens in the child node, with `"linger time"` being the first and the value being the second.
This should instead be something like:
```c++
		else if(!remove && child.Size() == 2 && child.Token(0) == "linger time")
			lingerTime = child.Value(1);
```

```c++
		else if(child.Size() > keyIndex && child.Token(keyIndex) == "linger time")
			child.PrintTrace("Invalid \"linger time\" specification.");
```
We generally don't discard something entirely just because it has too many tokoens. This is also looking at the wrong index for `"linger time"`.

I think the conditions should also be reordered and placed within a nested branch.
```suggestion
		else if(child.Token(0) == "linger time)
		{
			if(remove)
				lingerTime = -1;
			else if(child.Size() >= 2)
				lingerTime = child.Value(1);
			else
				child.PrintTrace("Invalid \"linger time\" specification.");
		}
```

Reply:
The linger time isn't always at index 0, so that wouldn't work. I see what you're getting at, and I'll try to simplify this code.

Reply:
I've applied your suggested change manually, except: 1) I added the missing `"` and 2) The error message is more specific.